man/lvmdiskscan.8
.TH LVMDISKSCAN 8 "LVM TOOLS" "Sistina Software UK" \" -*- nroff -*-
.SH NAME
lvmdiskscan \- scan for all devices visible to LVM2
.SH SYNOPSIS
.B lvmdiskscan
[\-d/\-\-debug] [\-h/\-?/\-\-help] 
[\-l/\-\-lvmpartition]
[\-v/\-\-verbose]
.SH DESCRIPTION
vgscan scans all SCSI, (E)IDE disks, multiple devices and a bunch
of other block devices in the system looking for LVM physical volumes.
The size reported is the real device size.
Define a filter in \fBlvm.conf\fP(5) to restrict 
the scan to avoid a CD ROM, for example.
.SH OPTIONS
See \fBlvm\fP for common options.
.TP
.I \-l, \-\-lvmpartition
Only reports Physical Volumes.
.SH SEE ALSO
.BR lvm (8), 
.BR lvm.conf (5),
.BR pvscan (8),
.BR vgscan (8)


debian/changelog
lvm2 (1.95.11-1) unstable; urgency=low

  * New upstream release.  (Closes: #171436)
  * Removed TODO and INTRO from debian/docs; added WHATS_NEW.
  * Remove vgcfgrestore.8 undocumented symlink.
  * Added a README.Debian, mentioning the device-mapper kernel module
    requirement that lvm2 has.  (Closes: #171674)
  * Get rid of debian/conffiles (debhelper's smart enough to figure that out).
  * debian/copyright fix to appease lintian.
  * Fix typo in tools/commands.h that caused /usr/sbin/; to be created.

 -- Andres Salomon <dilinger@mp3revolution.net>  Mon,  9 Dec 2002 02:51:02 -0400

lvm2 (1.95.10-2) unstable; urgency=low

  * Fix software raid problems by ensuring lvm init script runs after
    raidtools init script.  (Closes: #152569)

 -- Andres Salomon <dilinger@mp3revolution.net>  Tue,  3 Sep 2002 04:05:43 -0400

lvm2 (1.95.10-1) unstable; urgency=low

  * New upstream release (Beta 3.2).
  * Change all references to /dev/device-mapper/control to
    /dev/mapper/control.

 -- Andres Salomon <dilinger@mp3revolution.net>  Sun,  1 Sep 2002 18:55:12 -0400

lvm2 (0.95.05-3) unstable; urgency=low

  * Get rid of awk dependency in init script.  (Closes: #146257)

 -- Andres Salomon <dilinger@mp3revolution.net>  Sun, 12 May 2002 04:39:06 -0500

lvm2 (0.95.05-2) unstable; urgency=low

  * Use ${shlibs:Depends} in Depends.
  * Get rid of postinst/postrm scripts, use debhelper's init script instead.
  * Add Conflicts against lvm10, lvm-common.
  * Fix endian issues on big-endian machines.

 -- Andres Salomon <dilinger@mp3revolution.net>  Thu,  2 May 2002 23:53:53 -0500

lvm2 (0.95.05-1) unstable; urgency=low

  * New release (Beta2).

 -- Andres Salomon <dilinger@mp3revolution.net>  Thu, 25 Apr 2002 00:37:41 -0500

lvm2 (0.95.04cvs20020306-1) unstable; urgency=low

  * CVS updated.
  * Convert from debian native package.

 -- Andres Salomon <dilinger@mp3revolution.net>  Wed,  6 Mar 2002 00:43:21 -0500

lvm2 (0.95.04cvs20020304) unstable; urgency=low

  * CVS updated.
  * Enhance init script; create devmapper control device, etc.
  * Add dmsetup as a suggestion.
  * Add /etc/lvm/lvm.conf conffile.
  * Add undocumented(7) for the commands missing manpages.
  
 -- Andres Salomon <dilinger@mp3revolution.net>  Mon,  4 Mar 2002 04:51:26 -0500

lvm2 (0.95.02cvs20020220) unstable; urgency=low

  * Initial Release.

 -- Andres Salomon <dilinger@mp3revolution.net>  Wed, 20 Feb 2002 03:17:25 -0500



tools/lvmdiskscan.c
/*
 * Copyright (C) 2002-2004 Sistina Software, Inc. All rights reserved.
 * Copyright (C) 2004-2007 Red Hat, Inc. All rights reserved.
 *
 * This file is part of LVM2.
 *
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License v.2.1.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 */

/*
 * Changelog
 *
 *   05/02/2002 - First drop [HM]
 */

#include "tools.h"

int disks_found;
int parts_found;
int pv_disks_found;
int pv_parts_found;
int max_len;

static int _get_max_dev_name_len(struct dev_filter *filter)
{
	int len = 0;
	int maxlen = 0;
	struct dev_iter *iter;
	struct device *dev;

	if (!(iter = dev_iter_create(filter, 1))) {
		log_error("dev_iter_create failed");
		return 0;
	}

	/* Do scan */
	for (dev = dev_iter_get(iter); dev; dev = dev_iter_get(iter)) {
		len = strlen(dev_name(dev));
		if (len > maxlen)
			maxlen = len;
	}
	dev_iter_destroy(iter);

	return maxlen;
}

static void _count(struct device *dev, int *disks, int *parts)
{
	int c = dev_name(dev)[strlen(dev_name(dev)) - 1];

	if (!isdigit(c))
		(*disks)++;
	else
		(*parts)++;
}

static void _print(struct cmd_context *cmd, const struct device *dev,
		   uint64_t size, const char *what)
{
	log_print("%-*s [%15s] %s", max_len, dev_name(dev),
		  display_size(cmd, size), what ? : "");
}

static int _check_device(struct cmd_context *cmd, struct device *dev)
{
	char buffer;
	uint64_t size;

	if (!dev_open_readonly(dev))
		return_0;

	if (!dev_read(dev, UINT64_C(0), (size_t) 1, &buffer)) {
		stack;
		if (!dev_close(dev))
			stack;
		return 0;
	}
	if (!dev_get_size(dev, &size)) {
		log_error("Couldn't get size of \"%s\"", dev_name(dev));
		size = 0;
	}
	_print(cmd, dev, size, NULL);
	_count(dev, &disks_found, &parts_found);
	if (!dev_close(dev)) {
		log_error("dev_close on \"%s\" failed", dev_name(dev));
		return 0;
	}
	return 1;
}

int lvmdiskscan(struct cmd_context *cmd, int argc __attribute__((unused)),
		char **argv __attribute__((unused)))
{
	uint64_t size;
	struct dev_iter *iter;
	struct device *dev;
	struct label *label;

	/* initialise these here to avoid problems with the lvm shell */
	disks_found = 0;
	parts_found = 0;
	pv_disks_found = 0;
	pv_parts_found = 0;

	if (arg_count(cmd, lvmpartition_ARG))
		log_warn("WARNING: only considering LVM devices");

	max_len = _get_max_dev_name_len(cmd->full_filter);

	if (!(iter = dev_iter_create(cmd->full_filter, 0))) {
		log_error("dev_iter_create failed");
		return ECMD_FAILED;
	}

	/* Do scan */
	for (dev = dev_iter_get(iter); dev; dev = dev_iter_get(iter)) {
		/* Try if it is a PV first */
		if ((label_read(dev, &label, UINT64_C(0)))) {
			if (!dev_get_size(dev, &size)) {
				log_error("Couldn't get size of \"%s\"",
					  dev_name(dev));
				continue;
			}
			_print(cmd, dev, size, "LVM physical volume");
			_count(dev, &pv_disks_found, &pv_parts_found);
			continue;
		}
		/* If user just wants PVs we are done */
		if (arg_count(cmd, lvmpartition_ARG))
			continue;

		/* What other device is it? */
		if (!_check_device(cmd, dev))
			continue;
	}
	dev_iter_destroy(iter);

	/* Display totals */
	if (!arg_count(cmd, lvmpartition_ARG)) {
		log_print("%d disk%s",
			  disks_found, disks_found == 1 ? "" : "s");
		log_print("%d partition%s",
			  parts_found, parts_found == 1 ? "" : "s");
	}
	log_print("%d LVM physical volume whole disk%s",
		  pv_disks_found, pv_disks_found == 1 ? "" : "s");
	log_print("%d LVM physical volume%s",
		  pv_parts_found, pv_parts_found == 1 ? "" : "s");

	return ECMD_PROCESSED;
}
